artifact/tar/tar.hpp

```diff
@@ -0,0 +1,104 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_COMMON_TAR_HPP
+#define MENDER_COMMON_TAR_HPP
+
+#include <string>
+#include <vector>
+
+#include <config.h>
+
+#include <common/io.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+
+#ifdef MENDER_TAR_LIBARCHIVE
+#include <libarchive/wrapper.hpp>
+#endif
+
+namespace mender {
+namespace tar {
+
+using namespace std;
+
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace io = mender::common::io;
+
+
+enum ErrorCode {
+	NoError = 0,
+	TarReaderError,
+	TarShortReadError,
+	TarEntryError,
+	TarEOFError,
+};
+
+class ErrorCategoryClass : public std::error_category {
+public:
+	const char *name() const noexcept override;
+	string message(int code) const override;
+};
+
+extern const ErrorCategoryClass ErrorCategory;
+
+error::Error MakeError(ErrorCode code, const string &msg);
+
+class Entry : public io::Reader {
+private:
+	std::string name_;
+	uint64_t total_size_;
+
+	Reader &reader_;
+
+	// Reader data
+	size_t nr_bytes_read_ {0};
+
+public:
+	Entry(std::string name, uint64_t archive_size, Reader &reader) :
+		name_ {name},
+		total_size_ {archive_size},
+		reader_ {reader} {
+	}
+
+	std::string Name() {
+		return name_;
+	}
```

Comment:
Can we also get the size? We will need this for [MEN-6273](https://northerntech.atlassian.net/browse/MEN-6273).